### ISSUE TITLE
FIXME: Code improvement

### ISSUE THREAD

**Issue by Mai0313** [Mai0313/Chess-AI-Pytorch#22]
**Risk Level 3 - /home/runner/work/Chess-AI-Pytorch/Chess-AI-Pytorch/src/models/chess_mcts_module.py**

- The code changes in the `model_step` function introduce a potential bug. The `state`, `target_policy`, and `target_value` variables are unpacked from the `batch` tuple, but there is no guarantee that the tuple will always contain these three elements. This could lead to a runtime error if the tuple structure changes. It would be safer to add some error handling or validation to ensure the expected elements are present in the tuple.

- The `model_step` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `model_step` function could be made cleaner by using more descriptive variable names. For example, `pi_logits` could be renamed to `policy_logits` to improve clarity.

- The `model_step` function could be optimized for performance by using PyTorch's built-in loss functions instead of manually calculating the losses. This would simplify the code and potentially improve efficiency. Here's an example of how the code could be modified:

```python
state, target_policy, target_value = batch
pi_logits, value = self.forward(state)

policy_loss = F.cross_entropy(pi_logits, target_policy)
value_loss = F.mse_loss(value, target_value)
total_loss = policy_loss + value_loss

losses = {\"policy_loss\": policy_loss, \"value_loss\": value_loss, \"total_loss\": total_loss}

return losses, pi_logits, value
```

- The `training_step` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `validation_step` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `on_validation_epoch_end` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `test_step` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `on_test_epoch_end` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

- The `setup` function could benefit from improved readability. Consider breaking it down into smaller, more focused functions to improve code organization and maintainability.

---

**Risk Level 3 - /home/runner/work/Chess-AI-Pytorch/Chess-AI-Pytorch/src/models/components/alpha_zero.py**

The code changes in the pull request have a moderate risk to the code base.

---

**Risk Level 3 - /home/runner/work/Chess-AI-Pytorch/Chess-AI-Pytorch/setup_engine.py**

The code changes look fine overall, but there are a few improvements that can be made:

1. In the `execute_command` function, consider using the `subprocess.run` function instead of `subprocess.call` for better control and capturing of the command output.

2. In the `install_requirements` function, instead of using `subprocess.call` to install packages one by one, consider using `subprocess.run` with the `check=True` argument to install all packages at once. This will improve performance.

3. In the `setup_windows` function, consider using the `requests` library instead of `urllib.request` for downloading files. The `requests` library provides a more convenient and Pythonic way of handling HTTP requests.

4. In the `setup_linux` function, consider using the `requests` library instead of `urllib.request` for downloading files.

5. In the `get_model` function, consider using the `requests` library instead of `urllib.request` for downloading files.

6. Consider adding error handling and logging for any potential exceptions that may occur during the execution of the script.

Example code snippet for improvement #1:

```python
import subprocess


def execute_command(command, error_message):
    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    if result.returncode != 0:
        print(f\"{error_message}: {result.stderr}\")
    return result.returncode
```

Example code snippet for improvement #2:

```python
import subprocess


def install_requirements():
    console.print(Panel.fit(Text(\"Removing Cache...\", style=\"bold green\")))
    execute_command(\"pip cache purge\", \"There is no Cache to remove\")
    console.print(Panel.fit(Text(\"Installing Requirements ...\", style=\"bold green\")))
    with open(\"requirements.txt\") as f:
        packages = f.readlines()
    packages = [package.strip() for package in packages]
    result = subprocess.run([\"pip\", \"install\"] + packages, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
    if result.returncode != 0:
        console.print(Panel.fit(Text(\"Package installation failed\", style=\"bold red\")))

```

---
🐛📚🔧

---

#### Powered by [Code Review GPT](https://github.com/mattzcarey/code-review-gpt)

_Originally posted by @github-actions[bot] in https://github.com/Mai0313/Chess-AI-Pytorch/issues/21#issuecomment-1749796411_
